kafka/src/main/java/org/apache/calcite/adapter/kafka/KafkaStreamTable.java

```diff
@@ -0,0 +1,115 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.adapter.kafka;
+
+import org.apache.calcite.DataContext;
+import org.apache.calcite.config.CalciteConnectionConfig;
+import org.apache.calcite.linq4j.AbstractEnumerable;
+import org.apache.calcite.linq4j.Enumerable;
+import org.apache.calcite.linq4j.Enumerator;
+import org.apache.calcite.rel.RelCollations;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeFactory;
+import org.apache.calcite.schema.ScannableTable;
+import org.apache.calcite.schema.Schema;
+import org.apache.calcite.schema.Statistic;
+import org.apache.calcite.schema.Statistics;
+import org.apache.calcite.schema.StreamableTable;
+import org.apache.calcite.schema.Table;
+import org.apache.calcite.sql.SqlCall;
+import org.apache.calcite.sql.SqlNode;
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.ConsumerConfig;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+
+import com.google.common.collect.ImmutableList;
+
+import java.util.Collections;
+import java.util.Properties;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * A table which maps to a Kafka topic, currently only {@link KafkaStreamTable} is implemented as
+ * a STREAM table.
+ */
+public class KafkaStreamTable implements ScannableTable, StreamableTable {
+  final KafkaTableOptions tableOptions;
+
+  KafkaStreamTable(final KafkaTableOptions tableOptions) {
+    this.tableOptions = tableOptions;
+  }
+
+  @Override public Enumerable<Object[]> scan(final DataContext root) {
+    final AtomicBoolean cancelFlag = DataContext.Variable.CANCEL_FLAG.get(root);
+    return new AbstractEnumerable<Object[]>() {
+      public Enumerator<Object[]> enumerator() {
+        if (tableOptions.getConsumer() != null) {
+          return new KafkaMessageEnumerator(tableOptions.getConsumer(),
+              tableOptions.getRowConverter(), cancelFlag);
+        }
+
+        Properties consumerConfig = new Properties();
+        consumerConfig.put(ConsumerConfig.BOOTSTRAP_SERVERS_CONFIG,
+            tableOptions.getBootstrapServers());
+        //by default it's <byte[], byte[]>
+        consumerConfig.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG,
+            "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+        consumerConfig.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG,
+            "org.apache.kafka.common.serialization.ByteArrayDeserializer");
+
+        if (tableOptions.getConsumerParams() != null) {
+          consumerConfig.putAll(tableOptions.getConsumerParams());
+        }
+        Consumer consumer = new KafkaConsumer<>(consumerConfig);
```

Comment:
Can we subscribe as late as possible (inside Enumerator on first `next`) ?

Reply:
`prepare` would be the best place, `next` looks odd to me.

Reply:
Probably it is fine as is for now. It can be improved later-on.